sound/soc/sof/sof-audio.c

```diff
@@ -112,6 +112,13 @@ int sof_widget_free(struct snd_sof_dev *sdev, struct snd_sof_widget *swidget)
 	if (--swidget->use_count)
 		return 0;
 
+	if (swidget->id == snd_soc_dapm_scheduler)
+		goto free;
+
+	/* continue freeing the widget even if this fails */
+	err = sof_widget_free(sdev, swidget->pipe_widget);
+
+free:
```

Comment:
the goto is a bit ugly, no?

````c
err = 0;
if (widget->id != snd_soc_dapm_scheduler) {
    /* continue freeing the widget even if this fails */
    err = sof_widget_free(sdev, swidget->pipe_widget);
}
````

Reply:
fixed this and also moved it to free pipe widget after freeing the widget itself.